platform/iphone/os_iphone.h

```diff
@@ -127,7 +127,9 @@ class OSIPhone : public OS_Unix {
 
 	virtual int get_screen_dpi(int p_screen = -1) const;
 
+	void pencil_press(int p_idx, int p_x, int p_y, bool p_pressed, bool p_doubleclick);
 	void touch_press(int p_idx, int p_x, int p_y, bool p_pressed, bool p_doubleclick);
+	void pencil_drag(int p_idx, int p_prev_x, int p_prev_y, int p_x, int p_y, float p_force);
 	void touch_drag(int p_idx, int p_prev_x, int p_prev_y, int p_x, int p_y);
 	void touches_cancelled(int p_idx);
```

Comment:
Shouldn't there also be a `pencil_cancel`, that would be called on `touchesCancelled` method for stylus? 
As far as I understand pencil events mapped to different type of events (`MouseEvent`), so canceling should also be separate?

Reply:
That makes sense...adding now.

Reply:
I'm afraid not, since `touches_canceled` is cancelling `InputTouchEvent` and you need to cancel `InputMouseEvent`.
You can probably use `pencil_press` for it, but I would prefer separate function to handle cancelling in case we need to extend it in future without touching unrelated stuff.

Reply:
Ok, will add pencil_cancel.  Side note, clang format is failing, have never used clang before, is it something I add to VS studio code or as git hook?

Reply:
> is it something I add to VS studio code or as git hook?

I use `clang-format -i` with the list of files to run formatting on before committing just in case. I guess there should be a better and documented way to do this somewhere, but I don't use it :)

Reply:
ok.  I can't find a cancel mouse event...would it just be mouse event release?

Reply:
Yeah, that's probably it - touch cancel is simply sending pressed `false` for touch id.

Reply:
I'm trying to figure out the pencil_cancelled, but need extra info for index, position etc that touch_cancelled doesn't have.  Sorry I keep getting stuck, not sure how this should work.
void OSIPhone::pencil_cancelled(int p_idx) {
	Ref<InputEventMouseButton> ev;
	ev.instance();
	ev->set_button_index(1);
	ev->set_pressed(false);
	ev->set_position(Vector2(p_x, p_y));
	ev->set_global_position(Vector2(p_x, p_y));
	ev->set_doubleclick(p_doubleclick);
	//ev->set_pressure(p_force);
	//ev->set_tilt(p_tilt);
	perform_event(ev);
}

Reply:
I think you should be fine just calling `pencil_press(p_idx, -1, -1, false, false);` in `pencil_cancelled` for now, just like it's done in `touch_cancelled`.

Reply:
Thanks, does this look ok?

```- (void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event {
	NSArray *tlist = [event.allTouches allObjects];
	for (unsigned int i = 0; i < [tlist count]; i++) {
		if ([touches containsObject:[tlist objectAtIndex:i]]) {
			UITouch *touch = [tlist objectAtIndex:i];
			int tid = [self getTouchIDForTouch:touch];
			ERR_FAIL_COND(tid == -1);
			if (touch.type == UITouchTypeStylus) {
				OSIPhone::get_singleton()->pencil_press(tid, touchPoint.x * self.contentScaleFactor, touchPoint.y * self.contentScaleFactor, false, false);
			}
			else {
				OSIPhone::get_singleton()->touches_cancelled(tid);
			}
		}
	}
	[self clearTouches];```

Reply:
I meant that you should probably just implement a `prencil_cancelled` function like this:

```
void OSIPhone::prencil_cancelled(int p_idx) {
	pencil_press(p_idx, -1, -1, false, false);
}
```

And call it like this: `OSIPhone::get_singleton()->prencil_cancelled(tid)`